### PR TITLE
backend/azurerm: upgrading the SDK / support for proxies

### LINKED ISSUE #18425: Support MSI (like EC2 roles) authentication when using a remote backend?
I'm using an Azure server as my terraform control machine, to do my deployments.  MSI solves the authentication issue of running terraform, but I can't init the remote backend file unless I provide an explicit storage key.  

Is there a way to authenticate via MSI for terraform init when using a remote backend?

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR upgrades to v21.3.0 of github.com/Azure/azure-sdk-for-go, v10.15.4 of github.comAzure/go-autorest and vendors github.com/hashicorp/go-azure-helpers @ 0.1.1

This is the first in a couple of PR's to the Backend which allows us to support additional authentication modes for the backend, which will allow us to fix #18425 and #16763.

This PR _intentionally_ doesn't fix the locking issue tracked in #17046 (see more details below) - however it should be possible to fix that once this is merged (since this changeset's already quite big)

Tasks:
- [x] Upgrade Azure/go-autorest / Azure/azure-sdk-for-go
- [x] Threading through the new auth package
- [x] Proxy Support
- [x] Switching to use the Resources Profile
- [x] Switching to use the Storage Profile
- [x] Setting the User Agent
- [x] Additional Acceptance Tests coverage for Service Principal Client Secret / Access Key
- [x] Manual Testing

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
This test currently fails in master:

```
$ TF_ACC=1 envchain azurerm go test -v ./backend/remote-state/azure/ -run="TestBackendLocked"
=== RUN   TestBackendLocked
2018/11/20 21:37:58 [DEBUG] Could not lock as state blob did not exist, creating with empty state
--- FAIL: TestBackendLocked (136.20s)
    backend_test.go:120: creating resource group terraform-backend-testing-1587647403404359918
    backend_test.go:126: creating storage account tfbackendtesting4ahk
    backend_test.go:140: fetching access key for storage account
    backend_test.go:157: creating container terraform
    backend_test.go:77: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("/nAdA86fOV+5XQZPSyBW/GuXG4OSTamTbKneOw8ucg3pg3w8JYgRxPGcOxrJ/2IUgTTdekiJviJ65jAc25i3LQ=="), "container_name":cty.StringVal("terraform"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("tfbackendtesting4ahk")}}
    backend_test.go:84: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("/nAdA86fOV+5XQZPSyBW/GuXG4OSTamTbKneOw8ucg3pg3w8JYgRxPGcOxrJ/2IUgTTdekiJviJ65jAc25i3LQ=="), "container_name":cty.StringVal("terraform"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("tfbackendtesting4ahk")}}
    backend_test.go:91: TestBackend: testing state locking for *azure.Backend
    backend_test.go:92: TestBackend: testing state locking for *azure.Backend
    backend_test.go:92: could not unlock with the reported ID "d4dcc229-3c31-8f1c-3bce-341391bd10f4": failed to delete lock info from metadata: storage: service returned error: StatusCode=412, ErrorCode=LeaseIdMissing, ErrorMessage=There is currently a lease on the blob and no lease ID was specified in the request.
        RequestId:7763c56f-901e-0022-3f10-811362000000
        Time:2018-11-20T20:38:02.3230462Z, RequestInitiated=Tue, 20 Nov 2018 20:38:02 GMT, RequestId=7763c56f-901e-0022-3f10-811362000000, API Version=2016-05-31, QueryParameterName=, QueryParameterValue=
        Lock Info:
          ID:        d4dcc229-3c31-8f1c-3bce-341391bd10f4
          Path:      terraform/testState
          Operation: test
          Who:       clientA
          Version:   0.12.0
          Created:   2018-11-20 20:38:00.379197 &#43;0000 UTC
          Info:
    backend_test.go:174: destroying created resources
    backend_test.go:184: Azure resources destroyed
FAIL
FAIL	github.com/hashicorp/terraform/backend/remote-state/azure	137.146s
```

And also fails in this branch:

```
$ TF_ACC=1 envchain azurerm go test -v ./backend/remote-state/azure/ -run="TestBackendLocked"
=== RUN   TestBackendLocked
2018/11/20 21:35:17 Testing if Service Principal / Client Certificate is applicable for Authentication..
2018/11/20 21:35:17 Testing if Service Principal / Client Secret is applicable for Authentication..
2018/11/20 21:35:17 Using Service Principal / Client Secret for Authentication
2018/11/20 21:35:17 Creating Resource Group "acctestrg-backend-9xc2"
2018/11/20 21:35:20 Creating Storage Account "acctestsa9xc2" in Resource Group "acctestrg-backend-9xc2"
2018/11/20 21:35:40 fetching access key for storage account
2018/11/20 21:35:40 Creating Container "acctestcont" in Storage Account "acctestsa9xc2" (Resource Group "acctestrg-backend-9xc2")
2018/11/20 21:35:41 [DEBUG] Could not lock as state blob did not exist, creating with empty state
2018/11/20 21:35:46 [DEBUG] Deleting Resource Group "acctestrg-backend-9xc2"..
2018/11/20 21:35:47 [DEBUG] Waiting for deletion of Resource Group "acctestrg-backend-9xc2"..
--- FAIL: TestBackendLocked (123.95s)
    backend_test.go:75: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("acctestsa9xc2"), "access_key":cty.StringVal("S8wiuV5VHzlzWd36JpHZKeq1qkwh0cdOcbec+vmP/Z1ETV7e+Au4QI2jkyDPRiD53GWLAAKS6th8muxOT5F8mA=="), "container_name":cty.StringVal("acctestcont")}}
    backend_test.go:82: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("S8wiuV5VHzlzWd36JpHZKeq1qkwh0cdOcbec+vmP/Z1ETV7e+Au4QI2jkyDPRiD53GWLAAKS6th8muxOT5F8mA=="), "container_name":cty.StringVal("acctestcont"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("acctestsa9xc2")}}
    backend_test.go:89: TestBackend: testing state locking for *azure.Backend
    backend_test.go:90: TestBackend: testing state locking for *azure.Backend
    backend_test.go:90: could not unlock with the reported ID "a1e0bd92-5967-e906-ccff-dd5a04e590f0": failed to delete lock info from metadata: storage: service returned error: StatusCode=412, ErrorCode=LeaseIdMissing, ErrorMessage=There is currently a lease on the blob and no lease ID was specified in the request.
        RequestId:e7446ee4-101e-0088-8010-81dd44000000
        Time:2018-11-20T20:35:45.6724620Z, RequestInitiated=Tue, 20 Nov 2018 20:35:45 GMT, RequestId=e7446ee4-101e-0088-8010-81dd44000000, API Version=2016-05-31, QueryParameterName=, QueryParameterValue=
        Lock Info:
          ID:        a1e0bd92-5967-e906-ccff-dd5a04e590f0
          Path:      acctestcont/testState
          Operation: test
          Who:       clientA
          Version:   0.12.0
          Created:   2018-11-20 20:35:43.604863 &#43;0000 UTC
          Info:
FAIL
FAIL	github.com/hashicorp/terraform/backend/remote-state/azure	124.944s
```

The underlying issue is being tracked in #17046 - since this functionality is currently broken in master, I'm going to tackle this in the next PR, since this PR's already quite large.

File: backend/remote-state/azure/arm_client.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/2017-03-09/resources/mgmt/resources"
+	armStorage "github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2016-05-01/storage"
```
Comment (tombuildsstuff) [inline backend/remote-state/azure/arm_client.go]: we'll probably want to use [the Profile](https://github.com/Azure/azure-sdk-for-go/blob/v21.3.0/profiles/2017-03-09/storage/mgmt/storage/models.go) here to ensure compatibility with Azure Stack

File: backend/remote-state/azure/backend.go [hunk 2]
```diff
@@ -83,6 +77,9 @@ func New() backend.Backend {
 				Description: "The Tenant ID.",
 				DefaultFunc: schema.EnvDefaultFunc("ARM_TENANT_ID", ""),
 			},
+
+			// TODO: rename these fields
+			// TODO: support for custom resource manager endpoints
```
Comment (tombuildsstuff) [inline backend/remote-state/azure/backend.go]: these are TODO's for a follow up PR

**Comment by tombuildsstuff** [comment]
aside from the existing failing test, the tests otherwise looks fine thus far:

```
$ TF_ACC=1 envchain azurerm go test -v ./backend/remote-state/azure/ -timeout 6000s
=== RUN   TestBackend_impl
--- PASS: TestBackend_impl (0.00s)
=== RUN   TestBackendConfig
--- PASS: TestBackendConfig (0.00s)
    backend_test.go:27: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("QUNDRVNTX0tFWQ0K"), "container_name":cty.StringVal("tfcontainer"), "key":cty.StringVal("state"), "storage_account_name":cty.StringVal("tfaccount")}}
=== RUN   TestBackend
2018/11/20 21:59:38 Testing if Service Principal / Client Certificate is applicable for Authentication..
2018/11/20 21:59:38 Testing if Service Principal / Client Secret is applicable for Authentication..
2018/11/20 21:59:38 Using Service Principal / Client Secret for Authentication
2018/11/20 21:59:38 Creating Resource Group "acctestrg-backend-aizw"
2018/11/20 21:59:43 Creating Storage Account "acctestsaaizw" in Resource Group "acctestrg-backend-aizw"
2018/11/20 22:00:04 fetching access key for storage account
2018/11/20 22:00:04 Creating Container "acctestcont" in Storage Account "acctestsaaizw" (Resource Group "acctestrg-backend-aizw")
2018/11/20 22:00:06 [DEBUG] Could not lock as state blob did not exist, creating with empty state
2018/11/20 22:00:07 [DEBUG] Could not lock as state blob did not exist, creating with empty state
2018/11/20 22:00:11 [DEBUG] Could not lock as state blob did not exist, creating with empty state
2018/11/20 22:00:12 [DEBUG] Deleting Resource Group "acctestrg-backend-aizw"..
2018/11/20 22:00:13 [DEBUG] Waiting for deletion of Resource Group "acctestrg-backend-aizw"..
--- PASS: TestBackend (142.35s)
    backend_test.go:51: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"storage_account_name":cty.StringVal("acctestsaaizw"), "access_key":cty.StringVal("qsGUbLnXYz7k4Kc8vlZ0CCJa2knYZrEMWbsGdrIwsT9NNaCAszpoystpj1ECx+0QmE3UuWhl+SQztuRHsZmXqQ=="), "container_name":cty.StringVal("acctestcont"), "key":cty.StringVal("testState")}}
=== RUN   TestBackendLocked
2018/11/20 22:02:01 Testing if Service Principal / Client Certificate is applicable for Authentication..
2018/11/20 22:02:01 Testing if Service Principal / Client Secret is applicable for Authentication..
2018/11/20 22:02:01 Using Service Principal / Client Secret for Authentication
2018/11/20 22:02:01 Creating Resource Group "acctestrg-backend-v0mr"
2018/11/20 22:02:03 Creating Storage Account "acctestsav0mr" in Resource Group "acctestrg-backend-v0mr"
2018/11/20 22:02:24 fetching access key for storage account
2018/11/20 22:02:24 Creating Container "acctestcont" in Storage Account "acctestsav0mr" (Resource Group "acctestrg-backend-v0mr")
2018/11/20 22:02:25 [DEBUG] Could not lock as state blob did not exist, creating with empty state
2018/11/20 22:02:29 [DEBUG] Deleting Resource Group "acctestrg-backend-v0mr"..
2018/11/20 22:02:30 [DEBUG] Waiting for deletion of Resource Group "acctestrg-backend-v0mr"..
--- FAIL: TestBackendLocked (137.01s)
    backend_test.go:75: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("UyQld+5g8I6089UjwP+xKIiK8Cx7lieLUoNt5M1SzmEcpAWwOfc4IoXYHWpOnTRQ0sx5EMkKamWSnkmuXLoF8Q=="), "container_name":cty.StringVal("acctestcont"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("acctestsav0mr")}}
    backend_test.go:82: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("UyQld+5g8I6089UjwP+xKIiK8Cx7lieLUoNt5M1SzmEcpAWwOfc4IoXYHWpOnTRQ0sx5EMkKamWSnkmuXLoF8Q=="), "container_name":cty.StringVal("acctestcont"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("acctestsav0mr")}}
    backend_test.go:89: TestBackend: testing state locking for *azure.Backend
    backend_test.go:90: TestBackend: testing state locking for *azure.Backend
    backend_test.go:90: could not unlock with the reported ID "856d5d9c-0895-6105-c846-7e38412810ef": failed to delete lock info from metadata: storage: service returned error: StatusCode=412, ErrorCode=LeaseIdMissing, ErrorMessage=There is currently a lease on the blob and no lease ID was specified in the request.
        RequestId:9066d562-b01e-0032-7514-81a3e7000000
        Time:2018-11-20T21:02:29.4485247Z, RequestInitiated=Tue, 20 Nov 2018 21:02:28 GMT, RequestId=9066d562-b01e-0032-7514-81a3e7000000, API Version=2016-05-31, QueryParameterName=, QueryParameterValue=
        Lock Info:
          ID:        856d5d9c-0895-6105-c846-7e38412810ef
          Path:      acctestcont/testState
          Operation: test
          Who:       clientA
          Version:   0.12.0
          Created:   2018-11-20 21:02:27.404873 &#43;0000 UTC
          Info:
=== RUN   TestRemoteClient_impl
--- PASS: TestRemoteClient_impl (0.00s)
=== RUN   TestRemoteClient
2018/11/20 22:04:18 Testing if Service Principal / Client Certificate is applicable for Authentication..
2018/11/20 22:04:18 Testing if Service Principal / Client Secret is applicable for Authentication..
2018/11/20 22:04:18 Using Service Principal / Client Secret for Authentication
2018/11/20 22:04:18 Creating Resource Group "acctestrg-backend-nuu9"
2018/11/20 22:04:22 Creating Storage Account "acctestsanuu9" in Resource Group "acctestrg-backend-nuu9"
2018/11/20 22:04:42 fetching access key for storage account
2018/11/20 22:04:42 Creating Container "acctestcont" in Storage Account "acctestsanuu9" (Resource Group "acctestrg-backend-nuu9")
2018/11/20 22:04:44 [DEBUG] Deleting Resource Group "acctestrg-backend-nuu9"..
2018/11/20 22:04:45 [DEBUG] Waiting for deletion of Resource Group "acctestrg-backend-nuu9"..
--- PASS: TestRemoteClient (119.37s)
    client_test.go:32: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"container_name":cty.StringVal("acctestcont"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("acctestsanuu9"), "access_key":cty.StringVal("MdxFcarPJ0ag3Ud7JUnmxlGN0zBu5Nr19OUQFBdAZBsYHgXOS/a/QnmUljwb+A9mQsltD9YkYiJlZJR/8Kz2iA==")}}
=== RUN   TestRemoteClientLocks
2018/11/20 22:06:17 Testing if Service Principal / Client Certificate is applicable for Authentication..
2018/11/20 22:06:17 Testing if Service Principal / Client Secret is applicable for Authentication..
2018/11/20 22:06:17 Using Service Principal / Client Secret for Authentication
2018/11/20 22:06:17 Creating Resource Group "acctestrg-backend-7ayh"
2018/11/20 22:06:21 Creating Storage Account "acctestsa7ayh" in Resource Group "acctestrg-backend-7ayh"
2018/11/20 22:06:41 fetching access key for storage account
2018/11/20 22:06:41 Creating Container "acctestcont" in Storage Account "acctestsa7ayh" (Resource Group "acctestrg-backend-7ayh")
2018/11/20 22:06:42 [DEBUG] Could not lock as state blob did not exist, creating with empty state
2018/11/20 22:06:44 [DEBUG] Deleting Resource Group "acctestrg-backend-7ayh"..
2018/11/20 22:06:44 [DEBUG] Waiting for deletion of Resource Group "acctestrg-backend-7ayh"..
--- PASS: TestRemoteClientLocks (135.29s)
    client_test.go:61: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("W7pkVRoEIk+uWbG0hadkKkbTe5uQkYyZ5m58M4P4vVPz12mx67/ktFr5tzvs+Ahh78bEiBwOPNEjpOg3qIug2A=="), "container_name":cty.StringVal("acctestcont"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("acctestsa7ayh")}}
    client_test.go:68: TestBackendConfig on *azure.Backend with configs.synthBody{Filename:"<TestWrapConfig>", Values:map[string]cty.Value{"access_key":cty.StringVal("W7pkVRoEIk+uWbG0hadkKkbTe5uQkYyZ5m58M4P4vVPz12mx67/ktFr5tzvs+Ahh78bEiBwOPNEjpOg3qIug2A=="), "container_name":cty.StringVal("acctestcont"), "key":cty.StringVal("testState"), "storage_account_name":cty.StringVal("acctestsa7ayh")}}
=== RUN   TestPutMaintainsMetaData
2018/11/20 22:08:32 Testing if Service Principal / Client Certificate is applicable for Authentication..
2018/11/20 22:08:32 Testing if Service Principal / Client Secret is applicable for Authentication..
2018/11/20 22:08:32 Using Service Principal / Client Secret for Authentication
2018/11/20 22:08:32 Creating Resource Group "acctestrg-backend-ucok"
2018/11/20 22:08:35 Creating Storage Account "acctestsaucok" in Resource Group "acctestrg-backend-ucok"
2018/11/20 22:08:56 fetching access key for storage account
2018/11/20 22:08:56 Creating Container "acctestcont" in Storage Account "acctestsaucok" (Resource Group "acctestrg-backend-ucok")
2018/11/20 22:08:58 [DEBUG] Deleting Resource Group "acctestrg-backend-ucok"..
2018/11/20 22:08:58 [DEBUG] Waiting for deletion of Resource Group "acctestrg-backend-ucok"..
--- PASS: TestPutMaintainsMetaData (118.62s)
FAIL
FAIL	github.com/hashicorp/terraform/backend/remote-state/azure	652.786s
```

**Comment by tombuildsstuff** [comment]
Tests prior to updating the names:

screenshot 2018-11-21 at 16 04 53

File: backend/remote-state/azure/helpers_test.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func buildTestClient(t *testing.T, res resourceNames) *ArmClient {
+	subscriptionID := os.Getenv("ARM_SUBSCRIPTION_ID")
+	tenantID := os.Getenv("ARM_TENANT_ID")
+	clientID := os.Getenv("ARM_CLIENT_ID")
+	clientSecret := os.Getenv("ARM_CLIENT_SECRET")
+	environment := os.Getenv("ARM_ENVIRONMENT")
+
+	if subscriptionID == "" || tenantID == "" || clientID == "" || clientSecret == "" {
```
Comment (tombuildsstuff) [inline backend/remote-state/azure/helpers_test.go]: we probably want to check `ARM_LOCATION` is set here 🤔

**Comment by tombuildsstuff** [comment]
Now that we're testing both Access Key and Service Principal authentication explicitly - the failing test has been duplicated (which is intentional) - but this otherwise passes on Azure Public & Azure Germany:

screenshot 2018-11-21 at 18 15 54

The failing tests in question (which will be fixed in the follow up PR):

screenshot 2018-11-21 at 18 16 12

**Comment by radeksimko** [comment]
Also this certainly needs review from someone more familiar with Azure than me.

**Pull request merged**